torchtune/modules/loss/kd_losses.py

```diff
@@ -138,3 +237,164 @@ def forward(
+            Default: 8
+        ignore_index (int): Specifies a target value that is ignored and does not contribute to the input gradient.
+            The loss is divided over non-ignored targets.
+            Default: -100
+    """
+
+    def __init__(self, num_output_chunks: int = 8, ignore_index: int = -100):
+        super().__init__()
+        self.num_output_chunks = num_output_chunks
+        self.ignore_index = ignore_index
+        self.rkl_loss = ReverseKLLoss(ignore_index)
+
+    def forward(
+        self,
+        student_logits: List[torch.Tensor],
+        teacher_logits: List[torch.Tensor],
+        labels: torch.Tensor,
+    ) -> torch.Tensor:
+        """
+        Args:
+            student_logits (List[torch.Tensor]): List of chunked logits from student model of length
+                ``self.num_output_chunks``, where each chunk has shape
+                (batch_size, num_tokens / num_output_chunks, vocab_size).
+            teacher_logits (List[torch.Tensor]): List of chunked logits from teacher model of length
+                ``self.num_output_chunks``, where each chunk has shape
+                (batch_size, num_tokens / num_output_chunks, vocab_size).
+            labels (torch.Tensor): Ground truth labels of shape (batch_size, num_tokens).
+
+        Returns:
+            torch.Tensor: KL divergence loss of shape (1,).
+
+        Example:
+            >>> loss_fn = ReverseKLWithChunkedOutputLoss()
+            >>>
+            >>> h = torch.tensor([bsz, num_tokens, dim])
+            >>> output_chunks = [model.output(chunk) for chunk in h.chunk(num_chunks, dim=1)]
+            >>> teacher_chunks = [teacher_model.output(chunk) for chunk in h.chunk(num_chunks, dim=1)]
+            >>> labels = torch.tensor([bsz, num_tokens])
+            >>> loss = loss_fn(output_chunks, teacher_chunks, labels)
+        """
+
+        # reshape logits [(bsz, num_tokens/num_chunks, vocab)] -> [(bsz*num_tokens/num_chunks, vocab)]
+        teacher_logits = [
+            teacher_logits_chunk.reshape(-1, teacher_logits_chunk.size(-1))
+            for teacher_logits_chunk in teacher_logits
+        ]
+        student_logits = [
+            student_logits_chunk.reshape(-1, student_logits_chunk.size(-1))
+            for student_logits_chunk in student_logits
+        ]
+        mask = (labels != self.ignore_index).int()
+        # chunk and reshape labels (bsz, num_tokens, vocab) -> [(bsz*num_tokens/num_chunks, vocab)]
+        labels = [
+            target_chunk.reshape(-1)
+            for target_chunk in labels.chunk(self.num_output_chunks, dim=1)
+        ]
+
+        total_rkl_loss = 0.0
+        for student_chunk, teacher_chunk, label_chunk in zip(
+            student_logits, teacher_logits, labels
+        ):
+            total_rkl_loss += self.rkl_loss(
+                student_chunk, teacher_chunk, label_chunk, normalize=False
+            )
+
+        return total_rkl_loss / torch.sum(mask.view(-1), dim=0)
```

Comment:
I don't think our existing chunked forward KL loss even does this, but I wonder why we don't do the same check that `torch.sum(mask.view(-1), dim=0) != 0` that we're doing in the unchunked version? Couldn't we still potentially get division by zero here?

Reply:
We didn't add this check for the chunked forward KL either, mainly because if all the labels are ignore index, there's probably something wrong with the training data. Wondering if we should add this check in the training module to cover all the loss cases, maybe in a separate diff?